### PR TITLE
[Core] Support load and unload LoRA in api server

### PR DESCRIPTION
**Pull request by Jeffwan**
Address #6275 #3308 #5491 

Add the two methods in the apiserver to allow user to load/unload lora adapters explicitly.

- Update the `serving_engine.py` and add `load` and `unload` with request checking
- Update `api_server.py` and add api endpoint and request object
- Override LoraRequest `__eq__` and `__hash__`. We used to use `lora_init_id` as the identifier. but the tricky problem we meet is user may not able to persist the id on the client side every time, using model name is more straightforward. what's more, there's no need to load multiple same lora adapters. 

My proposal is to use `lora_name` as the identifier, the other good thing is we can use consistent names across instances. For example, let's say we have 2-3 instances, we want all of them load a specific adapter and unload a specific adapter. Using name is more consistent in such scenarios.

> Note: Community already have similar PRs like #3446. I tried to contact the PR author but it's kind of stale. The current PR covers more edge cases and verified in our environment.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/lora/request.py [hunk 1]
```diff
@@ -71,3 +70,21 @@ def local_path(self, value):
...
+    def __eq__(self, value: object) -> bool:
+        """
+        Overrides the equality method to compare LoRARequest
+        instances based on lora_name. This allows for identification
+        and comparison lora adapter across engines.
+        """
+        return isinstance(value,
+                          self.__class__) and self.lora_name == value.lora_name
+
+    def __hash__(self) -> int:
```
Comment (Yard1) [inline vllm/lora/request.py:83]: we should mark the dataclass as frozen if we want to use hash

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -165,6 +167,36 @@ async def create_embedding(request: EmbeddingRequest, raw_request: Request):
         return JSONResponse(content=generator.model_dump())
 
 
+@router.post("/v1/load_lora_adapter")
+async def load_lora_adapter(request: LoadLoraAdapterRequest):
```
Comment (Yard1) [inline vllm/entrypoints/openai/api_server.py:171]: we should allow operators to disable those methods and have them disabled by default; having user specify the path to the adapter themselves is inherently unsafe.

Comment (Jeffwan) [reply vllm/entrypoints/openai/api_server.py:171]: makes sense. let me add some feature gate for it

File: vllm/lora/request.py (hunk 1 above)
Comment (Jeffwan) [reply vllm/lora/request.py:83]: yeah. good catch. I will add `@dataclass(frozen=True)` then

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (Jeffwan) [reply vllm/entrypoints/openai/api_server.py:233]: I added a feature flag now and leave a TODO here.  I think once it's enabled, it's still has the risk. The key problem is if we want to support admin operations, then we have to distinguish the users. This is something not supported yet. In our system, we distinguish the user at proxy level. If the vLLM is public to users directly, then that's a problem

File: vllm/lora/request.py (hunk 1 above)
Comment (Jeffwan) [reply vllm/lora/request.py:83]:
```
vllm/lora/request.py:37: error: Property "lora_path" defined in "LoRARequest" is read-only  [misc]
vllm/lora/request.py:72: error: Property "lora_path" defined in "LoRARequest" is read-only  [misc]
```

I meet some issues by adding `frozen`, the problem is we have compatibility support for deprecated field. At this moment, there's no lora_name modifications, I think it should be safe. Once we remove the deprecated field, we can add `frozen` back

**Comment by Jeffwan** [comment]
@jgordley Yeah, I am still working on it. I was off last week and I will spend some time fixing the UTs. Thanks for offering the help, I will let you know if I meet problems.

File: vllm/entrypoints/openai/serving_engine.py [hunk 3]
```diff
@@ -403,3 +405,76 @@ def _get_decoded_token(logprob: Logprob,
...
+
+    async def load_lora_adapter(
+            self,
+            request: LoadLoraAdapterRequest) -> Union[ErrorResponse, str]:
+        error_check_ret = await self._check_load_lora_adapter_request(request)
+        if error_check_ret is not None:
+            return error_check_ret
+
+        lora_name, lora_path = request.lora_name, request.lora_path
+        unique_id = len(self.lora_requests) + 1
```
[the remark this answers is not in the document]
Comment (Jeffwan) [reply vllm/entrypoints/openai/serving_engine.py:460]: This is a known issue and another PR will fix this issue. Even we change to use a global counter, user can still pass the int id and override it, which means the counter may generate duplicate id next time. Technically, we should always use model name from user and id should be managed by the engine.

**Comment by Jeffwan** [comment]
@Yard1 can you help review the updated commits?

**Comment by simon-mo** [comment]
@jeejeelee can you help review this one? thank you!

**Comment by jeejeelee** [comment]
> @jeejeelee can you help review this one? thank you!

If there's no rush, I'll take some time to look at it this weekend.

**Comment by jeejeelee** [comment]
@Jeffwan  Thank you very much for your great work. I've left some comments，thanks~

**Comment by simon-mo** [comment]
@jeejeelee you might need to submit your review in github, as it is not showing here.

File: vllm/envs.py [hunk 4]
```diff
@@ -408,6 +409,12 @@ def get_default_config_root():
     # If set, vLLM will use Triton implementations of AWQ.
     "VLLM_USE_TRITON_AWQ":
     lambda: bool(int(os.getenv("VLLM_USE_TRITON_AWQ", "0"))),
+
+    # If set, allow loading or unloading lora adapters in runtime,
+    "VLLM_ALLOW_RUNTIME_LORA_LOADING":
+    lambda:
+    (os.environ.get("VLLM_ALLOW_RUNTIME_LORA_LOADING", "0").strip().lower() in
+     ("1", "true")),
```
[the remark this answers is not in the document]
Comment (Jeffwan) [reply vllm/envs.py:417]: sounds good

**Comment by Jeffwan** [comment]
I will address all the comments by tomorrow

File: tests/entrypoints/openai/test_serving_engine.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@pytest.mark.asyncio
+async def test_unload_lora_adapter_not_found():
+    serving_engine = await _async_serving_engine_init()
+    request = UnloadLoraAdapterRequest(lora_name="nonexistent_adapter")
+    response = await serving_engine.unload_lora_adapter(request)
+    assert isinstance(response, ErrorResponse)
+    assert response.err_type == "InvalidUserInput"
+    assert response.status_code == HTTPStatus.BAD_REQUEST
```
[the remark this answers is not in the document]
Comment (Jeffwan) [reply tests/entrypoints/openai/test_serving_engine.py:101]: I may forgot to push the latest commit, it should be addressed

File: vllm/entrypoints/openai/api_server.py [hunk 6]
```diff
@@ -325,9 +327,44 @@ async def stop_profile():
...
+                                status_code=response.code)
+
+        response = await openai_serving_completion.unload_lora_adapter(request)
+        if isinstance(response, ErrorResponse):
+            return JSONResponse(content=response.model_dump(),
+                                status_code=response.code)
+
+        return Response(status_code=200)
+
+
```
[the remark this answers is not in the document]
Comment (Jeffwan) [reply vllm/entrypoints/openai/api_server.py:363]: that's a good idea. Let me do it

[the remark this answers is not in the document]
Comment (Jeffwan) [reply vllm/entrypoints/openai/api_server.py:378]: updated

**Comment by Jeffwan** [comment]
image
All the code related CI failures have been fixed, current failures are from AMD test suite and primary reason is the docker hub rate limit.

**Comment by Jeffwan** [comment]
@jeejeelee Hi, I address all the comments and please take another look

File: docs/source/models/lora.rst [hunk 7]
```diff
@@ -107,3 +107,55 @@ The following is an example request
             "max_tokens": 7,
             "temperature": 0
         }' | jq
+
+
+Dynamically serving LoRA Adapters
+---------------------------------
+
+In addition to serving LoRA adapters at server startup, the vLLM server now supports dynamically loading and unloading
+LoRA adapters at runtime through dedicated API endpoints.This feature can be particularly useful when the flexibility
```
Comment (jeejeelee) [inline docs/source/models/lora.rst:116]: LoRA adapters at runtime through dedicated API endpoints.This feature can be particularly useful when the flexibility
```suggestion
LoRA adapters at runtime through dedicated API endpoints. This feature can be particularly useful when the flexibility
```

File: docs/source/models/lora.rst [hunk 8]
```diff
@@ -107,3 +107,55 @@ The following is an example request
...
+is set to `True`. When this option is enabled, the API server will log a warning to indicate that dynamic loading is active.
+
+.. code-block:: bash
+
+    export VLLM_ALLOW_RUNTIME_LORA_UPDATING=True
+
+
+Loading a LoRA Adapter:
+
+To dynamically load a LoRA adapter, send a POST request to the `/v1/load_lora_adapter`` endpoint with the necessary
```
Comment (jeejeelee) [inline docs/source/models/lora.rst:131]: To dynamically load a LoRA adapter, send a POST request to the `/v1/load_lora_adapter` endpoint with the necessary
```suggestion
To dynamically load a LoRA adapter, send a POST request to the `/v1/load_lora_adapter`` endpoint with the necessary
```

**Comment by jeejeelee** [comment]
@Jeffwan  Thank you very much for your work again. This is indeed a useful feature.  After making a few more NIT changes, I will give the green light. cc @simon-mo

**Comment by Jeffwan** [comment]
@jeejeelee doc update looks good to me. I accepted the suggestions

**Review by simon-mo — APPROVED** [review]
stamp given @jeejeelee's comment. thank @Jeffwan for the hardware and thank @jeejeelee for the shepherding!

File: vllm/entrypoints/openai/serving_engine.py [hunk 9]
```diff
@@ -78,6 +81,7 @@ def __init__(
 
         self.served_model_names = served_model_names
 
+        self.lora_id_counter = AtomicCounter(0)
```
Comment (njhill) [inline vllm/entrypoints/openai/serving_engine.py:84]: @Jeffwan this is not required here. asyncio operations all happen in the same thread. Can change this to be a simple int field.

Comment (Jeffwan) [reply vllm/entrypoints/openai/serving_engine.py:84]: @njhill yeah, that makes sense. Let me file a follow up PR to improve it.

**Pull request merged**